applications/nrf5340_audio/src/bluetooth/le_audio_cis_gateway.c

```diff
@@ -4,3 +4,399 @@
+	int ret;
+	struct event_t event;
+
+	event.event_source = EVT_SRC_LE_AUDIO;
+	event.le_audio_activity.le_audio_evt_type = LE_AUDIO_EVT_NOT_STREAMING;
+
+	ret = ctrl_events_put(&event);
+}
+
+static void stream_released_cb(struct bt_audio_stream *stream)
+{
+	LOG_INF("Audio Stream %p released", (void *)stream);
+}
+
+static struct bt_audio_stream_ops stream_ops = {
+	.configured = stream_configured_cb,
+	.qos_set = stream_qos_set_cb,
+	.enabled = stream_enabled_cb,
+	.started = stream_started_cb,
+	.metadata_updated = stream_metadata_updated_cb,
+	.disabled = stream_disabled_cb,
+	.stopped = stream_stopped_cb,
+	.released = stream_released_cb,
+};
+
+static void add_remote_sink(struct bt_audio_ep *ep, uint8_t index)
+{
+	LOG_INF("Sink #%u: ep %p", index, (void *)ep);
+	sinks[index] = ep;
+}
+
+static void add_remote_codec(struct bt_codec *codec, int index, uint8_t type)
+{
+	if (type != BT_AUDIO_SINK && type != BT_AUDIO_SOURCE) {
+		return;
+	}
+
+	if (index < CONFIG_BT_AUDIO_UNICAST_CLIENT_PAC_COUNT) {
+		remote_codecs[index] = codec;
+	}
+}
+
+static void discover_sink_cb(struct bt_conn *conn, struct bt_codec *codec, struct bt_audio_ep *ep,
+			     struct bt_audio_discover_params *params)
+{
+	int ret;
+
+	if (params->err != 0) {
+		LOG_ERR("Discovery failed: %d", params->err);
+		return;
+	}
+
+	if (codec != NULL) {
+		add_remote_codec(codec, params->num_caps, params->type);
+		return;
+	}
+
+	if (ep != NULL) {
+		if (params->type == BT_AUDIO_SINK) {
+			add_remote_sink(ep, params->num_eps);
+		} else {
+			LOG_ERR("Invalid param type: %u", params->type);
+		}
+
+		return;
+	}
+
+	LOG_INF("Discover complete: err %d", params->err);
+
+	(void)memset(params, 0, sizeof(*params));
+	ret = bt_audio_stream_config(default_conn, &audio_stream, sinks[0],
+				     &unicast_present_nrf5340.codec);
+	if (ret != 0) {
+		LOG_ERR("Could not configure stream");
+	}
+}
+
+static int device_found(uint8_t type, const uint8_t *data, uint8_t data_len,
+			const bt_addr_le_t *addr)
+{
+	int ret;
+	char addr_str[BT_ADDR_LE_STR_LEN];
+
+	bt_addr_le_to_str(addr, addr_str, sizeof(addr_str));
+	if ((data_len == DEVICE_NAME_PEER_LEN) &&
+	    (strncmp(DEVICE_NAME_PEER, data, DEVICE_NAME_PEER_LEN) == 0)) {
+		LOG_INF("device found");
+		bt_le_scan_stop();
+		ret = bt_conn_le_create(addr, BT_CONN_LE_CREATE_CONN, BT_LE_CONN_PARAM_MULTI,
+					&default_conn);
+		if (ret != 0) {
+			LOG_INF("Create conn to failed (%u)", ret);
+			start_scan();
+		}
+	}
+
+	return -ENOENT;
+}
+
+/** @brief  BLE parse advertisement package.
+ */
+static void ad_parse(struct net_buf_simple *p_ad, const bt_addr_le_t *addr)
+{
+	while (p_ad->len > 1) {
+		uint8_t len = net_buf_simple_pull_u8(p_ad);
+		uint8_t type;
+
+		/* Check for early termination */
+		if (len == 0) {
+			return;
+		}
+
+		if (len > p_ad->len) {
+			LOG_INF("AD malformed");
+			return;
+		}
+
+		type = net_buf_simple_pull_u8(p_ad);
+
+		if (device_found(type, p_ad->data, len - 1, addr) == 0) {
```

Comment:
device_found never returns 0?

Reply:
Once if device got an advertising packet from scan result, ad_parse will try to walk through the entire advertising packet by checking the data set one by one. If device_found() got the interested device and would like to connect, then just return.
If not, then ad_parse goes through entire advertising packet and leave.